keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,364 @@
+-->
+
+[etcd.io]: https://etcd.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[etcd-io/website]: https://github.com/etcd-io/website
+
+## Summary
+
+We are introducing the Kubernetes style feature gate framework into etcd to gate future feature enhancement behind a sequence of feature lifecyles. Users would be able to turn features on or off and query feature enablement of an etcd server in a consistent way. 
+
+## Motivation
+
+Currently any new enhancements to the etcd are typically added as [an experimental feature](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/features.md#adding-a-new-feature), with a configuration flag prefixed with "experimental", e.g. --experimental-feature-name. 
+
+When it is time to [graduate an experimental feature to stable](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/features.md#graduating-an-experimental-feature-to-stable), a new stable feature flag identical to the experimental feature flag but without the --experimental prefix is added to replace the old feature flag, which is a breaking change.
+
+We are proposing to add [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) in the similar way as Kubernetes does, so that can turn on or off any feature using the `--feature-gates` flag.
+
+Benefits from this enhancement includes:
+
+* graduating a feature is no longer a breaking change with new flags.
+
+* well-established feature lifecycle convention from Kubernetes.
+
+* a unified place to store all the feature enablement state in the code base for simplicity, and easy to refer to in guarding the implementation code.
+
+* easy to query if a feature is available for an ectd server.
+
+### Goals
+
+* Introduce the `--feature-gates` flag in etcd code base.
+
+* Introduce apis/methods to query the enablement state of a feature.
+
+* Establish lifecycle stages of a feature, and clear criteria of lifecycle progression.
+
+### Non-Goals
+
+* Handle the feature gate enablement on the cluster level. It would be designed in a separate KEP.
+
+* Migrate all existing `--experimental` features to feature gate. 
+  * While feature gate provides a [path to deprecate existing `--experimental` features]((#path-to-migrate-existing-experimental-features)), and we will migrate several `--experimental` during the feature gate implementation as examples, the completion of the KEP does not require migrating all existing `--experimental` features. It can be done as followup tasks.
+
+* Use feature gate as a mechanism to fix bugs.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+It would be easier for developers to add new features without worrying too much about breaking etcd, because the feature would be Alpha first and disabled by default, and new changes would be mostly gated behind the feature gate.
```

Comment:
That's not a user story, a user story would be:
1. contributor adds a feature, they add `--experimental-my-feature-flag`
2. In next minor release want to graduate a feature, to do that they need to add `--my-feature-flag` flag, but cannot remove the old flag to avoid breaking change
3. Another minor release happens, we finally remove the flag, however due X years passing contributor left the project so we forgot about the flag.

Reply:
updated